### DISCUSSION TITLE
Newton-Raphson residual postprocessing

### DISCUSSION THREAD

**Discussion by pbouda95** [thelfer/tfel#273] [General]
Hello,

Is there a elegant way to extract the `@Integrator` NR residual ? I would like store it for an afterward analysis ...
I thought about putting it inside a '@StateVariable' but maybe there is already a support for it.

Cheers,

**Comment by thelfer** [comment]
@pbouda95 You have access to the residual in the `@AdditionalConvergenceChecks` code block where the `error` variable is defined.

Thus, you can do something like that:

```c++
@AuxiliaryStateVariable real residual;

@Integrator{
....
}

@AdditionalConvergenceChecks{
  residual = error;
}
```

See `mfront --help-keyword=Implicit:@AdditionalConvergenceChecks` for details.

**Comment by pbouda95** [comment]
Thanks for this @thelfer ! However this would be the residual at the end of the timestep wouldn't it ? ... Is there a way to store its evolution ?

**Reply by thelfer** [reply]
@pbouda95 You can, but it's tricky. More precisely, you must take into account the fact that arrays in MFront must have fixed sized. But you can trick it as follows (warning, very untested):

```c++
@AuxiliaryStateVariables real residuals[100];

@InitiLocalVariables{
  std::fill(residuals.begin(), residuals.end(), real(0));
}

@AdditionalConvergenceChecks{
  const auto mindex = static_cast<unsigned short>(residual.size()-1);
   if(iter>mindex){
    // shift residual values
    if(mindex>1){
      for(unsigned short i=0; i != mindex-1; ++i){
        residual[i] = residual[i+1];
      }
    }
  }
  residuals[min(iter, mindex)] = error;
}
```

**Reply by pbouda95** [reply]
Ok I see the idea ! I will check this imediately
